### ISSUE TITLE
Add rate limit extension point

### ISSUE THREAD

**Issue by achrinza** [loopbackio/loopback-next#6409]
## Suggestion

Provide extension points, along with sensible default implementations, to enable request quota tracking.

## Use Cases

When creating an API, developers want to be able to control how often a user may be able to access an API over a given period. In some cases, they may want a simple `requests/unit-time` or a more complex `requests/unit-time with bursts` (e.g. AWS) measurement of metrics.

While it may be straightforward to implement a simple `IP address` or `UserProfile` REST quota tracker, this may be inflexible for some developers who may prefer more granular control.

These are some potential target user requirements:

- Granular control over metric increment trigger
  As a developer, I may choose to rate-limit part of my REST API, not the GraphQL endpoint, or base it off more complex conditions. I want a simple, well-defined interface to trigger an increment in the time-(in)sensitive quota tracking. This is also useful when developing other non-REST application.

- Pluggable quota trackers
  As a developer, I do not want to write my own quota tracker, especially if it's a simple `request/min` quota limit.

- Flexible choice of identifiers
  As a developer, I want to have control over how I identify a user of the API, even across API types. Whether it'd be `UserProfile`, `IP Address`, or some other source.

- Easy drop-in for simple use-cases
  As a developer, I should be able to drop in an IP address-based quota tracker for the entire API with minimal effort.

## Examples

This is an initial design proposal:

![image](https://user-images.githubusercontent.com/25147899/93709260-846e8a80-fb6f-11ea-88d1-89ee3197022f.png)

**Key information:**
- In creating a custom `hook`, the users are responsible for triggering an increment in the `Quota Tracker` by providing a `unified identifier` anywhere in the application, usually within the request-response lifecycle.
  - The default `hook` implementation could be a REST middleware that uses `IP Address` as an identifier, and can be configured to use any `tracker` through a `@config()` binding.
  - Another OOTB `hook` implementation could be a REST middleware that uses a well-defined ID property of `UserProfile` as an identifier, and can be configured in a similar way.
- The `unified identifier` is a `string` that can be derived from a single identifier such as an IP Address or UserProfile ID. It is the hook's responsibility to pass this identifier to the `Quota Tracker` upon trigger.
  - The `unified identifier` can also be derived from other disparate identifiers, and it is the developer's responsibility to derive a common `unified identifier`.
  - A `tracker` cannot see another `tracker`'s stored data. Hence, the `unified identifier` only needs to be unique within the same `tracker`.
- A `Quota` can contain additional properties that may aid a `QuotaTracker` decision making.
- The `Storage Adapters` is a simple, common interface that allows quick swapping of underlying infrastructure.

**Proposed `QuotaTracker` interface**

```typescript
export interface QuotaTracker {
  /**
   * A user-friendly display name for the quota tracker. It should be unique,
   * but not a must.
   */
  name: string;

  /**
   * Trigger a quota usage increment. Creates a new quota account if identifier is not found.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   */
  triggerQuotaUsage(identifier: string): void | Promise<void>;

  /**
   * Resets an existing quota tracker that's linked to the identifier.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   */
  resetQuota(identifier: string): void | Promise<void>;

  /**
   * Gets the total and remaining unit-quota for a quota account.
   * 
   * @remarks
   * The values returned should only be used for informational purposes, as
   * quota representation can be subjective and open to intepretation. To check
   * if the quota limit has been reached or exceeded, use `hasExceededQuota`
   * instead.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   */
  getQuota(identifier: string): Quota | Promise<Quota>;

  /**
   * Check if the quota for a quota account has been reached or exceeded.
   * 
   * @remarks
   * `undefined` may be returned if the quota can't be calculated for whatever
   * reason. It is up to the _quota enforcer_ to decide if the request should
   * be accepted or denied.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   */
  hasExceededQuota(identifier: string): boolean | undefined;
}
```

**Proposed `QuotaStorage` interface**

```typescript
export interface QuotaStorage {
  /**
   * A user-friendly display name for the quota storage. It should be unique,
   * but not a must.
   */
  name: string;

  /**
   * Creates a new quota.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   */
  createQuota(identifier: string, initialQuota: Quota): void | Promise<void>;

  /**
   * Gets the total and remaining unit-quota for a quota account.
   * 
   * @remarks
   * The values returned should only be used for informational purposes, as
   * quota representation can be subjective and open to intepretation. To check
   * if the quota limit has been reached or exceeded, use `hasExceededQuota`
   * instead.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   */
  getQuota(identifier: string): Quota | Promise<Quota>;

  /**
   * Sets a quota for a quota account.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   * @param newQuota - The new quota.
   */
  setQuota(identifier: string, newQuota: Quota): Quota | Promise<Quota>;

  /**
   * Checks if the quota account exists.
   * 
   * @param identifier - A globally-unique string to identify the quota account.
   */
  quotaExists(identifier: string): boolean;

  /**
   * Deletes a quota account.
   * 
   * @param identifier  - A globally-unique string to identify the quota account.
   */
  deleteQuota(identifier: string): void | Promise<void>;
}
```

**Proposed `Quota` interface**

```typescript
export interface Quota {
  total: number;
  remaining: number;
  units: string;
  [key: string]: unknown;
}
```

## Basic Use-Cases

For basic use-cases, we can provide a default `SimpleQuotaTracker` that will be bound as a `SINGLETON`, to keep track of triggers and the linked `identifier`, and routinely remove (or "expire") triggers from storage after a certain time has elapsed.

## Acceptance criteria

TBD - will be filled by the team.

**Comment by raymondfeng** [comment]
@achrinza Thank you for the well-written proposal! I think more people refer it as `rate limiting`. Are there existing Express middleware that can do the job?

**Comment by achrinza** [comment]
There's a few I've skimmed over such as [express-rate-limit](https://www.npmjs.com/package/express-rate-limit), but they are limited to REST. Since lb4 does more than typical REST, and since a user's quota should be measurable across different API types, I would lean towards having our own implementation. This would give us more flexibility to fit into our infrastructure.

I've also stumbled upon this standard, which may be useful:

https://tools.ietf.org/id/draft-polli-ratelimit-headers-01.html

I've not read it in full, just dropping it here for future reference.
